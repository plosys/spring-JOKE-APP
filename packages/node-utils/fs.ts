import { Dirent, Stats } from 'node:fs';
import { mkdir, opendir, readFile, stat, writeFile } from 'node:fs/promises';
import { dirname, extname, join } from 'node:path';

import { compareStrings } from '@appsemble/utils';
import parseJson from 'parse-json';
import sortKeys from 'sort-keys';
import { Promisable } from 'type-fest';
import { parse, stringify } from 'yaml';

import { AppsembleError } from './index.js';

/**
 * Test if the error is a NodeJS errno exception.
 *
 * @param error The value to check
 * @param code If specified, check if the code matches
 * @returns Whether or not the error is a NodeJS errno exception.
 */
export function isErrno(error: unknown, code?: string): error is NodeJS.ErrnoException {
  if (!error || typeof error !== 'object') {
    return false;
  }
  const err = error as NodeJS.ErrnoException;
  if (code) {
    return err.code === code;
  }
  return typeof err.code === 'string';
}

/**
 * Read the content of a JSON or YAML file.
 *
 * @param path The path to the file to read.
 * @returns A tuple of the parsed content and the content as a string.
 */
export async function readData<R>(path: URL | string): Promise<[R, string]> {
  let content: string;
  const ext = extname(String(path));
  try {
    content = await readFile(path, 'utf8');
  } catch {
    throw new AppsembleError(`Error reading file ${path}`);
  }
  if (ext !== '.yaml' && ext !== '.yml' && ext !== '.json') {
    throw new AppsembleError(`Unknown file extension: ${path}`);
  }
  try {
    return [
      ext === '.json' ? parseJson(content) : (parse(content, { maxAliasCount: 10_000 }) as R),
      content,
    ];
  } catch (error: unknown) {
    throw new AppsembleError(`Error parsing ${path}\n${(error as Error).message}`);
  }
}

interface WriteDataOptions {
  /**
   * If false, donâ€™t sort the object keys.
   */
  readonly sort?: boolean;

  /**
   * A comparison function to use for sorting keys.
   *
   * By default natural sorting will be used.
   */
  readonly compare?: ((a: string, b: string) => number) | null;
}

/**
 * Write data to a file serialized as JSON or YAML.
 *
 * If `prettier` is available, the content is also formatted.
 *
 * @param path The file path to write the data to.
 * @param data The data to write to the file.
 * @param options Additional options for processing the data.
 * @returns The formatted content.
 */
export async function writeData(
  path: string,
  data: unknown,
  { compare = compareStrings, sort = true }: WriteDataOptions = {},
): Promise<string> {
  const sorted = sort ? sortKeys(data, { deep: true, compare: compare || undefined }) : data;
  let buffer: string;
  try {
    const { default: prettier 