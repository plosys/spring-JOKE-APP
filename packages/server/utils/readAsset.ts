
import { readFile } from 'node:fs/promises';

export const assetDir = new URL('../assets/', import.meta.url);

/**
 * Read a file from the server assets directory.
 *
 * @param filename The path to the file to read, relative to the assets directory.
 *
 * param encoding If specified, decode the file using this encoding.
 * @returns The content of the specified file.
 */
export function readAsset(filename: string): Promise<Buffer>;
export function readAsset(filename: string, encoding: BufferEncoding): Promise<string>;
export function readAsset(filename: string, encoding?: BufferEncoding): Promise<Buffer | string> {
  return readFile(new URL(filename, assetDir), encoding);
}