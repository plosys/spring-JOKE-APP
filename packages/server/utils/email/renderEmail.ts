
import { has } from '@appsemble/utils';
import { InlineCode, Link, Parent, YAML } from 'mdast';
import rehypeDocument from 'rehype-document';
import rehypeStringify from 'rehype-stringify';
import frontmatter from 'remark-frontmatter';
import remarkParse from 'remark-parse';
import remarkRehype from 'remark-rehype';
import remarkStringify from 'remark-stringify';
import { unified } from 'unified';
import { visit } from 'unist-util-visit';
import { parse } from 'yaml';

const remark = unified()
  .use(remarkParse)
  .use(frontmatter)
  .use(remarkStringify, { emphasis: '_' })
  .freeze();

const rehype = unified().use(remarkRehype).use(rehypeDocument).use(rehypeStringify).freeze();

interface Email {
  html: string;
  subject: string;
  text: string;
}

/**
 * Render a markdown email.
 *
 * @param template The body of the template to render.
 * @param values Values to pass to the template for rendering.
 * @param sub The subject of the email to send. If omitted, this is extracted from the markdown
 * email body.
 * @returns An email object that may be sent.
 */
export async function renderEmail(
  template: string,
  values: Record<string, string>,
  sub?: string,
): Promise<Email> {
  let subject = sub;
  const mdast = remark.parse(template);

  function replace(match: string, key: string): string {
    if (has(values, key)) {
      return values[key];
    }
    throw new Error(`Unknown template value: ${key}`);
  }

  visit(mdast, 'link', (node: Link) => {
    // eslint-disable-next-line no-param-reassign
    node.url = node.url.replace(/\/{{(\w+)}}/, replace);
  });
  visit(mdast, 'inlineCode', (node: InlineCode, index, parent: Parent) => {
    parent.children.splice(index, 1, {
      type: 'text',
      value: node.value.replace(/{{(\w+)}}/, replace),
    });
  });

  if (!sub) {
    visit(mdast, 'yaml', (node: YAML, index, parent: Parent) => {
      ({ subject } = parse(node.value) as Email);
      parent.children.splice(index, 1);
    });
  }

  const text = remark.stringify(mdast);
  const hast = await rehype.run(mdast);
  const html = rehype.stringify(hast);
  return { html, subject, text };
}