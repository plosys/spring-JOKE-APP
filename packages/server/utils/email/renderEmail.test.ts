import { readdirSync } from 'node:fs';
import { readFile } from 'node:fs/promises';
import { parse } from 'node:path';

import { assetDir } from '../readAsset.js';
import { renderEmail } from './renderEmail.js';

const tests = {
  emailAdded: [
    { greeting: 'Hello User', url: 'https://example.com' },
    { greeting: 'Hello', url: 'https://example.com' },
  ],
  organizationInvite: [
    { greeting: 'Hello User', url: 'https://example.com', organization: 'winners' },
    { greeting: 'Hello', url: 'https://example.com', organization: 'winners' },
  ],
  resend: [
    { greeting: 'Hello User', url: 'https://example.com', name: 'The Appsemble Team' },
    { greeting: 'Hello', url: 'https://example.com', name: 'The Appsemble Team' },
  ],
  reset: [
    { greeting: 'Hello User', url: 'https://example.com', name: 'The Appsemble Team' },
    { greeting: 'Hello', url: 'https://example.com', name: 'The Appsemble Team' },
  ],
  welcome: [
    { greeting: 'Hello User', url: 'https://example.com' },
    { greeting: 'Hello', url: 'https://example.com' },
  ],
  welcomeMember: [
    { greeting: 'Hello User', url: 'https://example.com', name: 'Example App' },
    { greeting: 'Hello', url: 'https://example.com', name: 'Example App' },
  ],
  appMemberEmailChange: [
    { greeting: 'Hello User', url: 'https://example.com', name: 'Example App' },
    { greeting: 'Hello', url: 'https://example.com', name: 'Example App' },
  ],
  teamInvite: [{ teamName: 'Test Team', appName: 'Test App', url: 'https://example.com' }],
};

it.each(readdirSync(new URL('email/', assetDir)).map((f) => parse(f).name))(
  'should have tests for %s',
  (name) => {
    expect(tests).toHaveProperty(name);
    expect(tests[name as keyof typeof tests].length).toBeGreaterThan(0);
  },
);

describe.each(Object.entries(tests))('%s', (name, testValues: Record<string, any>[]) => {
  it.each(testValues)(`should render ${name} %#`, async (values) => {
    const template = await readFile(new URL(`email/${name}.md`, assetDir), 'utf8');
    const { html, text } = await renderEmail(template, values);
    expect(text).toMatchSnapshot('text');
    expect(html).toMatchSnapshot('html');
  });
});
