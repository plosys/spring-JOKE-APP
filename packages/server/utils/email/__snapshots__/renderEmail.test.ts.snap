// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appMemberEmailChange should render appMemberEmailChange 0: html 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello User,</p>
<p>You have changed your email address for Example App. Before you can use your account, we need to
verify your email address.</p>
<p>Please click <a href="https://example.com">here</a> to verify your email address.</p>
<p>Kind regards,</p>
<p><em>Example App</em></p>
</body>
</html>
"
`;

exports[`appMemberEmailChange should render appMemberEmailChange 0: text 1`] = `
"Hello User,

You have changed your email address for Example App. Before you can use your account, we need to
verify your email address.

Please click [here](https://example.com) to verify your email address.

Kind regards,

_Example App_
"
`;

exports[`appMemberEmailChange should render appMemberEmailChange 1: html 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello,</p>
<p>You have changed your email address for Example App. Before you can use your account, we need to
verify your email address.</p>
<p>Please click <a href="https://example.com">here</a> to verify your email address.</p>
<p>Kind regards,</p>
<p><em>Example App</em></p>
</body>
</html>
"
`;

exports[`appMemberEmailChange should render appMemberEmailChange 1: text 1`] = `
"Hello,

You have changed your email address for Example App. Before you can use your account, we need to
verify your email address.

Please click [here](https://example.com) to verify your email address.

Kind regards,

_Example App_
"
`;

exports[`emailAdded should render emailAdded 0: html 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello User,</p>
<p>You have registered a new email address to your account. In order to make full use of this email, we
need to verify your email address.</p>
<p>Please click <a href="https://example.com">this link</a> to verify your email address.</p>
<p><strong>Didn’t request this email?</strong></p>
<p>No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will be done with this email address.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`emailAdded should render emailAdded 0: text 1`] = `
"Hello User,

You have registered a new email address to your account. In order to make full use of this email, we
need to verify your email address.

Please click [this link](https://example.com) to verify your email address.

**Didn’t request this email?**

No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will be done with this email address.

Kind regards,

_The Appsemble Team_
"
`;

exports[`emailAdded should render emailAdded 1: html 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello,</p>
<p>You have registered a new email address to your account. In order to make full use of this email, we
need to verify your email address.</p>
<p>Please click <a href="https://example.com">this link</a> to verify your email address.</p>
<p><strong>Didn’t request this email?</strong></p>
<p>No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will be done with this email address.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`emailAdded should render emailAdded 1: text 1`] = `
"Hello,

You have registered a new email address to your account. In order to make full use of this email, we
need to verify your email address.

Please click [this link](https://example.com) to verify your email address.

**Didn’t request this email?**

No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will be done with this email address.

Kind regards,

_The Appsemble Team_
"
`;

exports[`organizationInvite should render organizationInvite 0: html 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello User,</p>
<p>You have been invited to join the organization <strong>winners</strong>.</p>
<p>Joining an organization allows you to view private applications as well as create and edit apps
belonging to an organization.</p>
<p>You can accept or decline this invitation by following <a href="https://example.com">this link</a>.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`organizationInvite should render organizationInvite 0: text 1`] = `
"Hello User,

You have been invited to join the organization **winners**.

Joining an organization allows you to view private applications as well as create and edit apps
belonging to an organization.

You can accept or decline this invitation by following [this link](https://example.com).

Kind regards,

_The Appsemble Team_
"
`;

exports[`organizationInvite should render organizationInvite 1: html 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello,</p>
<p>You have been invited to join the organization <strong>winners</strong>.</p>
<p>Joining an organization allows you to view private applications as well as create and edit apps
belonging to an organization.</p>
<p>You can accept or decline this invitation by following <a href="https://example.com">this link</a>.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`organizationInvite should render organizationInvite 1: text 1`] = `
"Hello,

You have been invited to join the organization **winners**.

Joining an organization allows you to view private applications as well as create and edit apps
belonging to an organization.

You can accept or decline this invitation by following [this link](https://example.com).

Kind regards,

_The Appsemble Team_
"
`;

exports[`resend should render resend 0: html 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello User,</p>
<p>Someone has requested to resend the verification key for your account. Before you can use your
account, we need to verify your email address.</p>
<p>Please click <a href="https://example.com">this link</a> below to verify your email address</p>
<p><strong>Didn’t request this email?</strong></p>
<p>No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will happen to your account.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`resend should render resend 0: text 1`] = `
"Hello User,

Someone has requested to resend the verification key for your account. Before you can use your
account, we need to verify your email address.

Please click [this link](https://example.com) below to verify your email address

**Didn’t request this email?**

No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will happen to your account.

Kind regards,

_The Appsemble Team_
"
`;

exports[`resend should render resend 1: html 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello,</p>
<p>Someone has requested to resend the verification key for your account. Before you can use your
account, we need to verify your email address.</p>
<p>Please click <a href="https://example.com">this link</a> below to verify your email address</p>
<p><strong>Didn’t request this email?</strong></p>
<p>No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will happen to your account.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`resend should render resend 1: text 1`] = `
"Hello,

Someone has requested to resend the verification key for your account. Before you can use your
account, we need to verify your email address.

Please click [this link](https://example.com) below to verify your email address

**Didn’t request this email?**

No worries! Your address may have been entered by mistake. If you ignore or delete this email,
nothing will happen to your account.

Kind regards,

_The Appsemble Team_
"
`;

exports[`reset should render reset 0: html 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello User,</p>
<p>A password reset has been requested for you account. Please click <a href="https://example.com">here</a> to reset your
password.</p>
<p><strong>Didn’t request this email?</strong></p>
<p>Your email address may have been entered by mistake. This email can be ignored safely.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`reset should render reset 0: text 1`] = `
"Hello User,

A password reset has been requested for you account. Please click [here](https://example.com) to reset your
password.

**Didn’t request this email?**

Your email address may have been entered by mistake. This email can be ignored safely.

Kind regards,

_The Appsemble Team_
"
`;

exports[`reset should render reset 1: html 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello,</p>
<p>A password reset has been requested for you account. Please click <a href="https://example.com">here</a> to reset your
password.</p>
<p><strong>Didn’t request this email?</strong></p>
<p>Your email address may have been entered by mistake. This email can be ignored safely.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`reset should render reset 1: text 1`] = `
"Hello,

A password reset has been requested for you account. Please click [here](https://example.com) to reset your
password.

**Didn’t request this email?**

Your email address may have been entered by mistake. This email can be ignored safely.

Kind regards,

_The Appsemble Team_
"
`;

exports[`teamInvite should render teamInvite 0: html 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello,</p>
<p>You have been invited to join the team <strong>Test Team</strong> for the app <strong>Test App</strong>.</p>
<p>You can accept this invitation by following <a href="https://example.com">this link</a>.</p>
<p>Kind regards,</p>
<p><em>The Test App Team</em></p>
</body>
</html>
"
`;

exports[`teamInvite should render teamInvite 0: text 1`] = `
"Hello,

You have been invited to join the team **Test Team** for the app **Test App**.

You can accept this invitation by following [this link](https://example.com).

Kind regards,

_The Test App Team_
"
`;

exports[`welcome should render welcome 0: html 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello User,</p>
<p>Thank you for registering your account. Before you can use your account, we need to verify your
email address.</p>
<p>Please click <a href="https://example.com">here</a> to verify your email address.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`welcome should render welcome 0: text 1`] = `
"Hello User,

Thank you for registering your account. Before you can use your account, we need to verify your
email address.

Please click [here](https://example.com) to verify your email address.

Kind regards,

_The Appsemble Team_
"
`;

exports[`welcome should render welcome 1: html 1`] = `
"<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello,</p>
<p>Thank you for registering your account. Before you can use your account, we need to verify your
email address.</p>
<p>Please click <a href="https://example.com">here</a> to verify your email address.</p>
<p>Kind regards,</p>
<p><em>The Appsemble Team</em></p>
</body>
</html>
"
`;

exports[`welcome should render welcome 1: text 1`] = `
"Hello,

Thank you for registering your account. Before you can use your account, we need to verify your
email address.

Please click [here](https://example.com) to verify your email address.

Kind regards,

_The Appsemble Team_
"
`;

exports[`welcomeMember shoul